File: src/动态规划/template/DP类/树形dp/P1352_没有上司的舞会/Main.cpp
#include <iostream>
#include <cstring>
#include <algorithm>
using namespace std;

const int N=6010;
int head[N],to[N],ne[N],idx;
void add(int a,int b){
  to[++idx]=b,ne[idx]=head[a],head[a]=idx;
}
int n,w[N],fa[N];
int f[N][2]; //0不选,1选

void dfs(int u){
  f[u][1]=w[u];
  for(int i=head[u];i;i=ne[i]){
    int v=to[i];
    dfs(v);
    f[u][0]+=max(f[v][0],f[v][1]);
    f[u][1]+=f[v][0];
  }
}
int main(){
  cin>>n;
  for(int i=1;i<=n;i++) cin>>w[i];
  for(int i=0,a,b;i<n-1;i++){
    cin>>a>>b;
    add(b,a);
    fa[a]=true;
  }
  int root=1;
  while(fa[root]) root++;
  dfs(root);
  cout<<max(f[root][0],f[root][1]);
}

File: src/动态规划/template/DP类/树形dp/P1352_没有上司的舞会/Main.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(100000)

def dfs(u):
    f[u][1] = w[u]  # 选u的快乐指数,赋初值
    for v in a[u]:
        dfs(v)
        f[u][0] += max(f[v][0], f[v][1])
        f[u][1] += f[v][0]


n = int(input())
w = [0] * 6010
a = [[] for i in range(n + 5)]  # 邻接表存图
fa = [0] * 6009
f = [[0] * 2 for _ in range(6008)]
for i in range(1, n + 1):
    w[i] = int(input())
for i in range(n - 1):
    x, y = map(int, input().split())
    a[y].append(x)
    fa[x] = 1  # 有父节点

root = 1
while fa[root]:
    root += 1
dfs(root)
print(max(f[root][0], f[root][1]))